### PR TITLE
Basic archive assemblies

### PR DESCRIPTION
**Pull request by est77**
Basically a way to include geometry, materials and lights from some project into other projects.
Uses of this could be supporting Maya's XGen (or other similar instancing tools), stand-ins, ...

The PR also contains a minimalistic implementation of procedurals that could be used later to support  Alembic for example.

### CODE REVIEW

**Comment by dictoon** [comment]
Fantastic move, thanks!

File: sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed [hunk 1]
```diff
@@ -0,0 +1,130 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- File generated by appleseed.so version 1.6.0-beta (SSE 4.2). -->
+<project format_revision="13">
+    <search_paths>
+        <search_path>
+            ../_assets/textures
+        </search_path>
+        <search_path>
+            ../bsdf/disneybrdf
```
Comment (dictoon) [inline sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed]: I don't think it's a good idea for test scenes to reference each others. What are you using from that directory?

File: sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            <parameter name="turbidity_multiplier" value="1.0" />
+        </environment_edf>
+        <environment_shader name="environment_shader" model="edf_environment_shader">
+            <parameter name="environment_edf" value="environment_edf" />
+        </environment_shader>
+        <environment name="environment" model="generic_environment">
+            <parameter name="environment_edf" value="environment_edf" />
+        </environment>
+        <assembly name="assembly" model="archive_assembly">
+<<<<<<< HEAD
```
Comment (dictoon) [inline sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed]: Incomplete merge.

File: src/appleseed.python/bindproject.cpp [hunk 3]
```diff
@@ -305,7 +305,9 @@ void bind_project()
     bpy::enum_<ProjectFileReader::Options>("ProjectFileReaderOptions")
         .value("Defaults", ProjectFileReader::Defaults)
         .value("OmitReadingMeshFiles", ProjectFileReader::OmitReadingMeshFiles)
-        .value("OmitProjectFileUpdate", ProjectFileReader::OmitProjectFileUpdate);
+        .value("OmitProjectFileUpdate", ProjectFileReader::OmitProjectFileUpdate)
+        .value("OmitSearchPaths", ProjectFileReader::OmitSearchPaths)
```
Comment (dictoon) [inline src/appleseed.python/bindproject.cpp]: Intuitively, without reading the rest of the code, I'm not sure what `OmitSearchPaths` means or achieves. Maybe the name could be made more descriptive.

Edit: Now that I've read the code, maybe `OmitAddingSearchPaths` would be clearer?

File: src/appleseed/renderer/modeling/project/projectfilereader.cpp [hunk 4]
```diff
@@ -3057,6 +3060,13 @@ auto_release_ptr<Project> ProjectFileReader::read(
     if (!xerces_context.is_initialized())
         return auto_release_ptr<Project>(0);
 
+    if ((options & OmitProjectSchemaValidation) == false && schema_filepath == 0)
+    {
+        RENDERER_LOG_ERROR(
+            "Project schema validation enabled, but no schema filepath provided.");
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: Start log message with a lowercase letter.

File: src/appleseed/renderer/modeling/project/projectfilereader.cpp [hunk 5]
```diff
@@ -3082,6 +3092,68 @@ auto_release_ptr<Project> ProjectFileReader::read(
...
+    assert(archive_filepath);
+
+    XercesCContext xerces_context(global_logger());
+    if (!xerces_context.is_initialized())
+        return auto_release_ptr<Assembly>(0);
+
+    if ((options & OmitProjectSchemaValidation) == false && schema_filepath == 0)
+    {
+        RENDERER_LOG_ERROR(
+            "Archive schema validation enabled, but no schema filepath provided.");
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: Start log message with a lowercase letter.

File: src/appleseed/renderer/modeling/project/projectfilereader.cpp [hunk 6]
```diff
@@ -3082,6 +3092,68 @@ auto_release_ptr<Project> ProjectFileReader::read(
...
+        false,
+        event_counters,
+        stopwatch.get_seconds());
+
+    if (event_counters.has_errors())
+        return auto_release_ptr<Assembly>(0);
+
+    if (project->get_scene())
+    {
+        if (Assembly* assembly = project->get_scene()->assemblies().get_by_name("assembly"))
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: Is it on purpose that the name "assembly" is hardcoded here?

File: src/appleseed/renderer/modeling/project/projectfilereader.cpp [hunk 7]
```diff
@@ -3136,13 +3218,24 @@ auto_release_ptr<Project> ProjectFileReader::load_project_file(
...
-    parser->setProperty(
-        XMLUni::fgXercesSchemaExternalNoNameSpaceSchemaLocation,
-        const_cast<void*>(
-            static_cast<const void*>(
-                transcode(schema_filepath).c_str())));
+
+    if ((options & OmitProjectSchemaValidation) == false)
+    {
+        assert(schema_filepath);
+        parser->setFeature(XMLUni::fgSAX2CoreValidation, true);         // report all validation errors
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: Align comment.

File: src/appleseed/renderer/modeling/project/projectfilereader.h [hunk 8]
```diff
@@ -52,60 +54,71 @@ class APPLESEED_DLLSYMBOL ProjectFileReader
   public:
     enum Options
     {
-        Defaults                = 0,        // none of the flags below
-        OmitReadingMeshFiles    = 1 << 0,   // do not read mesh files from disk
-        OmitProjectFileUpdate   = 1 << 1    // do not update the project file format to the latest revision
+        Defaults                    = 0,        // none of the flags below
+        OmitReadingMeshFiles        = 1 << 0,   // do not read mesh files from disk
+        OmitProjectFileUpdate       = 1 << 1,   // do not update the project file format to the latest revision
+        OmitSearchPaths             = 1 << 2,   // do not read search paths from the project.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.h]: Remove period at the end of the end-of-line comment.

File: src/appleseed/renderer/modeling/project/projectfilewriter.cpp [hunk 9]
```diff
@@ -319,6 +320,19 @@ namespace
             if (strcmp(assembly.get_model(), AssemblyFactory().get_model()) != 0)
                 element.add_attribute("model", assembly.get_model());
 
+            // Don't write the contents of the assembly if they were
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilewriter.cpp]: *if it was*

File: src/appleseed/renderer/modeling/scene/archiveassembly.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "archiveassembly.h"
+
+// Standard headers.
+#include<string>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.cpp]: Add space.

File: src/appleseed/renderer/modeling/scene/archiveassembly.cpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+{
+
+//
+// ArchiveAssembly class implementation.
+//
+
+namespace
+{
+
+const char* Model = "archive_assembly";
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.cpp]: Indent contents of anonymous namespaces.

File: src/appleseed/renderer/modeling/scene/archiveassembly.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+bool ArchiveAssembly::expand_contents(
+    const Project&          project,
+    const Assembly*         parent,
+    IAbortSwitch*           abort_switch)
+{
+    if (!m_archive_opened)
+    {
+        // Establish and store the qualified path to the archive project.
+        const SearchPaths& search_paths = project.search_paths();
+        std::string filepath = search_paths.qualify(m_params.get_required<string>("filename", ""));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.cpp]: Remove redundant namespace qualifiers.

File: src/appleseed/renderer/modeling/scene/archiveassembly.h [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_RENDERER_MODELING_SCENE_ARCHIVE_ASSEMBLY_H
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.h]: `APPLESEED_RENDERER_MODELING_SCENE_ARCHIVEASSEMBLY_H`

File: src/appleseed/renderer/modeling/scene/archiveassembly.h [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+#define APPLESEED_RENDERER_MODELING_SCENE_ARCHIVE_ASSEMBLY_H
+
+// appleseed.renderer headers.
+#include "renderer/global/globaltypes.h"
+#include "renderer/modeling/scene/proceduralassembly.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.h]: Reorder.

File: src/appleseed/renderer/modeling/scene/archiveassembly.h [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+namespace foundation    { class StringArray; }
+namespace foundation    { class StringDictionary; }
+namespace renderer      { class ParamArray; }
+namespace renderer      { class Project; }
+
+namespace renderer
+{
+
+//
+// An assembly is either entirely self-contained, or it references colors,
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.h]: Is that comment still relevant / up-to-date?

File: src/appleseed/renderer/modeling/scene/archiveassembly.h [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    ArchiveAssembly(
+        const char*                 name,
+        const ParamArray&           params);
+
+    bool m_archive_opened;
+};
+
+
+//
+// ArchiveAssemblyFactory factory.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.h]: I think you meant *ArchiveAssembly factory*.

File: src/appleseed/renderer/modeling/scene/archiveassembly.h [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+//
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/archiveassembly.h]: Remove redundant blank line.

File: src/appleseed/renderer/modeling/scene/proceduralassembly.cpp [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "proceduralassembly.h"
+
+// Standard headers.
+#include<string>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/proceduralassembly.cpp]: Add space.

File: src/appleseed/renderer/modeling/scene/proceduralassembly.cpp [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+// appleseed.renderer headers.
+#include "renderer/utility/paramarray.h"
+
+using namespace foundation;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/proceduralassembly.cpp]: Remove redundant `using` clauses.

File: src/appleseed/renderer/modeling/scene/proceduralassembly.h [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_RENDERER_MODELING_SCENE_PROCEDURAL_ASSEMBLY_H
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/proceduralassembly.h]: `APPLESEED_RENDERER_MODELING_SCENE_PROCEDURALASSEMBLY_H`

File: src/appleseed/renderer/modeling/scene/proceduralassembly.h [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+namespace renderer      { class ParamArray; }
+
+namespace renderer
+{
+
+//
+// An assembly is either entirely self-contained, or it references colors,
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/proceduralassembly.h]: Fix/update comment as appropriate.

File: src/appleseed/renderer/modeling/scene/scene.cpp [hunk 22]
```diff
@@ -286,6 +287,49 @@ void Scene::on_render_end(const Project& project)
...
+        const Assembly*         parent,
+        IAbortSwitch*           abort_switch)
+    {
+        ProceduralAssembly *proc_assembly =
+            dynamic_cast<ProceduralAssembly*>(&assembly);
+
+        if (proc_assembly)
+        {
+            const bool ok = proc_assembly->expand_contents(project, parent, abort_switch);
+            if (!ok)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/scene.cpp]: I guess we can get rid of all the `ok` variables here?

File: src/appleseed.python/bindproject.cpp (hunk 3 above)
Comment (est77) [inline src/appleseed.python/bindproject.cpp]: We want to use the search paths in the "parent" project, not the ones in the project we are importing.

What happens is that we copy the search path of the parent project into the new project and then pass OmitSearchPaths to ignore the search paths in the project we are reading.

File: sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed (hunk 1 above)
Comment (est77) [inline sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed]: I think textures and geometry

Comment (dictoon) [inline sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed]: In that case, the shared assets should be moved to the `_assets` directory.

File: src/appleseed/renderer/modeling/project/projectfilereader.cpp (hunk 6 above)
Comment (est77) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: I think this is our convention. Studio, Gafferseed and appleseed-maya all name the main assembly "assembly". I'm not sure about appleseed-max.
Hardcoding the name simplifies things and I don't think it's too restrictive.
Nearly all archives will be generated by exporting from Maya / Max or Gaffer.

Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfilereader.cpp]: Right, but it's only a convention. I agree this is enough for now, I just wanted to make sure that wasn't a leftover.

File: src/appleseed.python/bindproject.cpp (hunk 3 above)
Comment (est77) [inline src/appleseed.python/bindproject.cpp]: I think that OmitSearchPaths is more accurate. It describes better what this option does.
Not only it does not add searchpaths from the project file, it also doesn't set the root path in the project.

File: sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed (hunk 1 above)
Comment (est77) [inline sandbox/tests/test scenes/archive assemblies/01 - archive assemblies.appleseed]: Fixed

**Pull request merged**